### PR TITLE
trying to see if pulling data from data sources 3 and 4 is doable

### PR DESCRIPTION
**Pull request by felipe74590**

### CODE REVIEW

File: web_scraper.py [hunk 1]
```diff
@@ -151,21 +157,64 @@ def gather_events_data_source_heyaustin(url: str, events_list=None) -> list[Even
...
-    # events = events_list
-    # print(events_soup)
-    # for event in events_soup:
-    #     title = event.find("div", class_="headline").text.strip()
-    #     print(title)
+    #The following are the relevant data keys withing the feed entry
+    #dict_keys(['title', 'link', 'tags', 'summary', 'summary_detail'])
+    feed = feedparser.parse(url)
+
+    for entry in feed.entries:
```
Comment (felipe74590) [inline web_scraper.py]: Issue: Location and datetime appear in the summary value but in a huge string. I have made some attempts to pull those values from the string but it has been in consistent.

File: web_scraper.py [hunk 2]
```diff
@@ -151,21 +157,64 @@ def gather_events_data_source_heyaustin(url: str, events_list=None) -> list[Even
...
+            page.wait_for_selector('.grid-flow-row-dense')
+            element_handle = page.query_selector('div.grid-flow-row-dense')
+            inner_html = element_handle.inner_html()
+            soup = BeautifulSoup(inner_html, 'html.parser')
+            events = soup.find_all('a')
+        
+        #TODO: Solve issue that doesn't allow event.find('div', class_='headline') to work
+        # Same with event.find('div', _class='location')
+        # Issue with date, the current function only shows time, this data source might be inefficient
+            for event in events:
```
Comment (felipe74590) [inline web_scraper.py]: Code comments address the main reason it has been a challange to pull the needed values from the data using playwright with beauftifulsoup.

File: web_scraper.py (hunk 1 above)
Comment (felipe74590) [reply web_scraper.py]: I also tried to use playwrite to pull html but ran into this error https://docs.google.com/document/d/1dVt6G6oJl-SEkP3hFiUR6q-VRSbvPmLUtyqU_Y-qsoo/edit that I couldn't get past.

Comment (bbelderbos) [reply web_scraper.py]: damn that rss output is not useful :( 
trying playwright then ...

File: main.py [hunk 3]
```diff
@@ -5,14 +5,16 @@
...
+    gather_events_data_source_atxorg,
     gather_events_data_source_do512,
     gather_events_data_source_heyaustin,
+    gather_events_data_atx_culture
 )
 
 source_one = config("SOURCE1")
 source_two = config("SOURCE2")
 source_three = config("SOURCE3")
+source_four = config("SOURCE4")
```
Comment (bbelderbos) [inline main.py]: As discussed, please put constants back.

File: main.py [hunk 4]
```diff
@@ -25,10 +27,12 @@
...
+            data = gather_events_data_source_atxorg(source_three)
         case "source4":
+            data = gather_events_data_atx_culture(source_four)
+        case "source5":
             data = get_predict_api_events()
         case _ :
             print("Invalid input")
 
-    add_events_to_db(data)
+    # add_events_to_db(data)
```
Comment (bbelderbos) [inline main.py]: could add boolean arg to write to database

File: web_scraper.py (hunk 1 above)
Comment (bbelderbos) [reply web_scraper.py]: Source 3 = `gather_events_data_source_atxorg`

This works for me. Well kind of, I see the invites twice but it's a start:

```
    with sync_playwright() as p:
        with p.chromium.launch(headless=False) as browser:
            page = browser.new_page()
            page.goto(url)
            breakpoint()
```

Debug:

```
(Pdb) for event in page.query_selector_all("div.contents"): event.inner_text()
'\n                Fantastic Fest\n                \n                \n                     1911 Aldrich St., Austin, TX 78723\n                \n            '
'\n                "Grit to Greatness" group exhibition\n                \n                \n                     2801 S I-35 Frontage Rd, Austin, TX 78741\n                \n            '
'\n                Black Film Geniuses: "The Thing That Comes" Movie Premiere\n                \n                \n                     6406 N IH-35 Frontage Road, Austin, TX 78752\n                \n            '
'\n                Free Tango Monday\n                \n                \n                     603 Sabine St, Austin, TX 78701\n                \n            '
'\n                Free “La Juventud” Youth Dance Classes for kids\n                \n                \n                     209 Pedernales Street, Austin, TX 78702\n                \n            '
'\n                Julia C. Butridge Gallery Exhibit: Adorn: Stories of Found Memory\n                \n                \n                     1110 Barton Springs Rd., Austin, TX 78704\n                \n            '
'\n                Julia C. Butridge Gallery Exhibit: Freedom House\n                \n                \n                     1110 Barton Springs Rd., Austin, TX 78704\n                \n            '
'\n                Kick the Bucket Podcast Staged Reading\n                \n                \n                     4100 Jackson Ave., Austin, TX 78731\n                \n            '
'\n                Legacies of the Treaty of Guadalupe Hidalgo Exhibition\n                \n                \n                     1800 N. Congress Ave., Austin, TX 78701\n                \n            '
'\n                Pumpkin Nights\n                \n                \n                     10621 Pioneer Farms Dr., Austin, TX 78754\n                \n            '
'\n                Round 22 : Plastic Bag Lion Dance\n                \n                \n                     916 Springdale Rd., Austin, TX 78702\n                \n            '
'\n                Texas Artists / Texas Music\n                \n                \n                     1110 Barton Springs Rd., Austin, TX 78704\n                \n            '
'FANTASTIC FEST\n\n Alamo Drafthouse Cinema - Mueller\n\n09/19/2024 - 09/26/2024\n\n Quick View\nof Fantastic Fest'
'"GRIT TO GREATNESS" GROUP EXHIBITION\n\n Art for the People Gallery\n\n08/23/2024 - 10/25/2024\n\n Quick View\nof "Grit to Greatness" group exhibition'
'BLACK FILM GENIUSES: "THE THING THAT COMES" MOVIE PREMIERE\n\n AFS Cinema & Event Hall\n\n09/23/2024\n\n Quick View\nof Black Film Geniuses: "The Thing That Comes" Movie Premiere'
'FREE TANGO MONDAY\n\n Vaquero Taquero (Downtown)\n\nRecurring weekly on Monday until 09/23/2024\n\n Quick View\nof Free Tango Monday'
'FREE “LA JUVENTUD” YOUTH DANCE CLASSES FOR KIDS\n\n Esquina Tango\n\nDates vary between 09/09/2024 - 10/10/2024\n\n Quick View\nof Free “La Juventud” Youth Dance Classes for kids'
'JULIA C. BUTRIDGE GALLERY EXHIBIT: ADORN: STORIES OF FOUND MEMORY\n\n Dougherty Arts Center\n\nRecurring weekly on Monday, Tuesday, Wednesday, Thursday, Friday, Saturday until 10/05/2024\n\n Quick View\nof Julia C. Butridge Gallery Exhibit: Adorn: Stories of Found Memory'
'JULIA C. BUTRIDGE GALLERY EXHIBIT: FREEDOM HOUSE\n\n Dougherty Arts Center\n\nDates vary between 08/10/2024 - 10/05/2024\n\n Quick View\nof Julia C. Butridge Gallery Exhibit: Freedom House'
'KICK THE BUCKET PODCAST STAGED READING\n\n The Westminster\n\nDates vary between 09/23/2024 - 09/25/2024\n\n Quick View\nof Kick the Bucket Podcast Staged Reading'
'LEGACIES OF THE TREATY OF GUADALUPE HIDALGO EXHIBITION\n\n Bullock Texas State History Museum\n\n02/16/2024 - 02/16/2025\n\n Quick View\nof Legacies of the Treaty of Guadalupe Hidalgo Exhibition'
'PUMPKIN NIGHTS\n\n Pioneer Farms\n\n09/19/2024 - 10/31/2024\n\n Quick View\nof Pumpkin Nights'
'ROUND 22 : PLASTIC BAG LION DANCE\n\n Canopy\n\n08/03/2024 - 11/02/2024\n\n Quick View\nof Round 22 : Plastic Bag Lion Dance'
'TEXAS ARTISTS / TEXAS MUSIC\n\n Dougherty Arts Center\n\nRecurring weekly on Monday, Tuesday, Wednesday, Thursday, Friday, Saturday until 10/26/2024\n\n Quick View\nof Texas Artists / Texas Music'
```

We could also go into each page to get more granular info, which does seem available to bs4:

```
import requests
import bs4

response = requests.get('https://www.austintexas.org/event/free-tango-monday/382722/')
soup = bs4.BeautifulSoup(response.text, 'html.parser')
print(soup.select('h1')[0].getText())
```

File: main.py (hunk 4 above)
Comment (bbelderbos) [reply main.py]: weird now this works again for me:

```
    args = parser.parse_args()

    print(f"{args.source=}")

    match args.source:
```

File: pyproject.toml [hunk 5]
```diff
@@ -6,6 +6,8 @@ readme = "README.md"
 requires-python = ">=3.12"
 dependencies = [
     "beautifulsoup4>=4.12.3",
+    "feedparser>=6.0.11",
```
Comment (bbelderbos) [inline pyproject.toml]: can remove now

File: web_scraper.py [hunk 6]
```diff
@@ -151,21 +157,64 @@ def gather_events_data_source_heyaustin(url: str, events_list=None) -> list[Even
...
-    # if events_list is None:
-    #     events_list = []
-    # events = events_list
-    # print(events_soup)
-    # for event in events_soup:
-    #     title = event.find("div", class_="headline").text.strip()
-    #     print(title)
+    #The following are the relevant data keys withing the feed entry
+    #dict_keys(['title', 'link', 'tags', 'summary', 'summary_detail'])
+    feed = feedparser.parse(url)
```
Comment (bbelderbos) [inline web_scraper.py]: let's drop this

File: web_scraper.py (hunk 2 above)
Comment (bbelderbos) [reply web_scraper.py]: image

Problem here is that you only get the time, but you could scrape the date from the outer div?

Comment (bbelderbos) [reply web_scraper.py]: This works for me:

```
(Pdb) for event in page.query_selector_all("div.event-post")[:5]: event.inner_text()
"EDITOR'S PICK\nHarry Ransom Center presents “Visualizing the Environment: Ansel Adams and His Legacy”\nHarry Ransom Center\n10:00 am"
"EDITOR'S PICK\nFantastic Fest\nAlamo Drafthouse Cinema South Lamar\n11:00 am"
'Live Oak Brewing Company presents 27th Annual Oaktoberfest Party\nLive Oak Brewing Company\n5:00 pm'
'Jessica Fontenot presents "200 Days Art Show"\nThe Cathedral\n6:00 pm'
'Mark Chesnutt in concert\nHaute Spot\n6:00 pm'
```

Comment (bbelderbos) [reply web_scraper.py]: This seems pretty reliable:

```
(Pdb) for event in page.query_selector_all("div.event-post")[:5]: event.inner_text().splitlines()
["EDITOR'S PICK", 'Harry Ransom Center presents “Visualizing the Environment: Ansel Adams and His Legacy”', 'Harry Ransom Center', '10:00 am']
["EDITOR'S PICK", 'Fantastic Fest', 'Alamo Drafthouse Cinema South Lamar', '11:00 am']
['Live Oak Brewing Company presents 27th Annual Oaktoberfest Party', 'Live Oak Brewing Company', '5:00 pm']
['Jessica Fontenot presents "200 Days Art Show"', 'The Cathedral', '6:00 pm']
['Mark Chesnutt in concert', 'Haute Spot', '6:00 pm']
(Pdb) for event in page.query_selector_all("div.event-post")[:10]: event.inner_text().splitlines()
["EDITOR'S PICK", 'Harry Ransom Center presents “Visualizing the Environment: Ansel Adams and His Legacy”', 'Harry Ransom Center', '10:00 am']
["EDITOR'S PICK", 'Fantastic Fest', 'Alamo Drafthouse Cinema South Lamar', '11:00 am']
['Live Oak Brewing Company presents 27th Annual Oaktoberfest Party', 'Live Oak Brewing Company', '5:00 pm']
['Jessica Fontenot presents "200 Days Art Show"', 'The Cathedral', '6:00 pm']
['Mark Chesnutt in concert', 'Haute Spot', '6:00 pm']
['Pioneer Farms presents Pumpkin Nights', 'Jourdan-Bachman Pioneer Farms', '6:00 pm']
['Esther’s Follies Theater presents Bigly Battle at the Ballot Box', "Esther's Follies", '7:00 pm']
['Moontower Comedy presents Therapy Gecko Live: The Lizard Agenda Tour', 'Stateside at the Paramount', '7:00 pm']
['Chromeo and The Midnight in concert', 'ACL Live at The Moody Theater', '7:00 pm']
['Cap City Comedy Club presents Arnez J', 'Capitol City Comedy Club', '7:00 pm']
```

Comment (bbelderbos) [reply web_scraper.py]: (you would only have to ditch the editor's pick and you have a same length list you can unpack)

Also nice:

```
(Pdb) len(page.query_selector_all("div.event-post"))
289
```

File: main.py (hunk 3 above)
Comment (bbelderbos) [reply main.py]: I did here: https://github.com/felipe74590/atx-events/pull/16 (changing rss feed to regular page)

File: web_scraper.py [hunk 7]
```diff
@@ -1,11 +1,15 @@
 
 from datetime import datetime
 from typing import NamedTuple
+import feedparser
+import re
```
Comment (bbelderbos) [inline web_scraper.py]: can drop now

File: web_scraper.py (hunk 1 above)
Comment (bbelderbos) [reply web_scraper.py]: committed here: https://github.com/felipe74590/atx-events/pull/16

File: web_scraper.py (hunk 2 above)
Comment (bbelderbos) [reply web_scraper.py]: can you take a look at this if this works for you? the challenge was to both parse the dates and grids to match them up:

https://github.com/felipe74590/atx-events/pull/16

Comment (felipe74590) [reply web_scraper.py]: Thanks this helped alot, just some for clean up and I was able to get everything from this source.

Comment (bbelderbos) [reply web_scraper.py]: Excellent, glad to hear!

File: src/db_helper.py [hunk 8]
```diff
@@ -2,7 +2,7 @@
 from sqlmodel import Field, Session, SQLModel, create_engine, select
 
 sqlite_file_name = config("DATABASE")
-sqlite_url = f"sqlite:///{sqlite_file_name}"
+sqlite_url = sqlite_file_name
```
Comment (bbelderbos) [inline src/db_helper.py]: can remove extra variable

File: src/events_api.py [hunk 9]
```diff
@@ -4,11 +4,11 @@
 import requests
 from decouple import config
 
-from web_scraper import Event
+from src.web_scraper import Event
 
 base_url = config("PREDICT_API")
 access_token = config("EVENTS_HQ_TOKEN")
-LIMIT = 50 #max events per page
+MAX_EVENTS_LIMIT = 50
```
Comment (bbelderbos) [inline src/events_api.py:11]: nice, more readable now

File: src/web_scraper.py [hunk 10]
```diff
@@ -151,21 +151,96 @@ def gather_events_data_source_heyaustin(url: str, events_list=None) -> list[Even
     return events
 
 
-def gather_events_data_source_atxmap(url: str, events_list=None) -> list[Event]:
+def gather_events_data_source_atxorg(url: str, events_list=None) -> list[Event]:
```
Comment (bbelderbos) [inline src/web_scraper.py:155]: could also add type hint to `events_list`

File: src/web_scraper.py [hunk 11]
```diff
@@ -151,21 +151,96 @@ def gather_events_data_source_heyaustin(url: str, events_list=None) -> list[Even
...
-    # if events_list is None:
-    #     events_list = []
-    # events = events_list
-    # print(events_soup)
-    # for event in events_soup:
-    #     title = event.find("div", class_="headline").text.strip()
-    #     print(title)
+    #The following are the relevant data keys withing the feed entry
+    #dict_keys(['title', 'link', 'tags', 'summary', 'summary_detail'])
+    with sync_playwright() as p, p.chromium.launch(headless=False) as browser:
```
Comment (bbelderbos) [inline src/web_scraper.py]: maybe nicer/ more readable to nest them like this?

```
    with sync_playwright() as p:
        with p.chromium.launch(headless=HEADLESS) as browser:
            yield browser
```

(and make `HEADLESS` configurable)

File: src/web_scraper.py [hunk 12]
```diff
@@ -151,21 +151,96 @@ def gather_events_data_source_heyaustin(url: str, events_list=None) -> list[Even
...
+        event_grids = page.query_selector_all("div.grid-flow-row-dense")
+
+        event_dates = [event.inner_text() for event in event_dates]
+
+        for date, grid in zip(event_dates, event_grids, strict=False):
+            for link in grid.query_selector_all("a"):
+                if link.get_attribute("href") is not None:
+                    event_link = link.get_attribute("href")
+
+            date_only = date.split("\n")
```
Comment (bbelderbos) [inline src/web_scraper.py]: or `splitlines()`

File: src/web_scraper.py [hunk 13]
```diff
@@ -151,21 +151,96 @@ def gather_events_data_source_heyaustin(url: str, events_list=None) -> list[Even
...
+
+            for event in grid.query_selector_all("div.event-post"):
```
Comment (bbelderbos) [inline src/web_scraper.py]: ok glad you could use this

**Review by bbelderbos — APPROVED** [review]
Great work, keep going!

**Pull request merged**